FD2 packed binary commands

tt-aho opened the pull request:
### Ticket
- https://github.com/tenstorrent/tt-metal/issues/9012

### Problem description
- FD2 was not packing binaries

### What's changed
- New prefetch/dispatch commands to handle large packed binaries

### Checklist
- [x] Post commit CI passes
- [x] Model regression CI testing passes (if applicable)
- [x] New/Existing tests provide coverage for changes

pgkeller committed:
#9012: FD2: prefetch packed read command

Step towards packing kernel binaries

pgkeller committed:
#9012: Dispatch packed write large command

Variation of packed write to handle "large" blocks as used by kernel dispatch

pgkeller committed:
#9012: Add new prefetch/dispatch to fd2 tests

pgkeller commented on tt_metal/impl/dispatch/kernels/cq_common.hpp:
```diff
@@ -9,8 +9,6 @@
 #include "debug/ring_buffer.h"
 
-#define L1_NOC_ALIGNMENT 16 // XXXXX is the defined elsewhere?
-
```
beautiful!

pgkeller commented on tt_metal/impl/dispatch/kernels/cq_prefetch.cpp:
```diff
@@ -74,6 +74,20 @@ static uint32_t block_next_start_addr[cmddat_q_blocks];
+constexpr uint32_t l1_to_local_cache_copy_chunk = 6;
+
+// Currently capping the same as dispatch
+// TODO: Should be its own max
```
this should be fixed before submitting imho

pgkeller commented on tt_metal/impl/dispatch/kernels/cq_prefetch.cpp:
```diff
@@ -613,35 +624,29 @@ uint32_t process_relay_paged_cmd(uint32_t cmd_ptr,
+        InterleavedPow2AddrGen<true> addr_gen;
         addr_gen.bank_base_address = base_addr;
-        addr_gen.page_size = page_size; // XXXX confirm mult goes away on GS or just replace
+        addr_gen.log_base_2_of_page_size = log_page_size; // XXXX confirm mult goes away on GS or just replace
```
oh, nice.  can remove the XXX comment, right?

pgkeller commented on tt_metal/impl/dispatch/kernels/cq_prefetch.cpp:
```diff
@@ -903,6 +925,84 @@ static uint32_t process_relay_inline_exec_buf_cmd(uint32_t& cmd_ptr,
+// That means this command is stateful, incorrect use will be...bad
+// NOTE: this routine assumes we're sending a command header and that is LESS THAN A PAGE
+static uint32_t process_relay_inline_noflush_exec_buf_cmd(uint32_t& cmd_ptr,
+                                                 uint32_t& dispatch_data_ptr) {
```
I think I should have named these "process_exec_buf_etc_cmd", care to fix if you agree?
I guess this has its own implementation because the size is known so the looping is simpler (less overhead), right?
can you add a comment about why this routine exists (how the command can get split by reading from dram)

pgkeller commented on tt_metal/impl/dispatch/kernels/cq_prefetch.cpp:
```diff
@@ -903,6 +925,84 @@ static uint32_t process_relay_inline_exec_buf_cmd(uint32_t& cmd_ptr,
+    return stride;
+}
+
+uint32_t process_relay_paged_packed_exec_buf_cmd(uint32_t& cmd_ptr,
```
can you add a comment about why this exists (vs the non exec_buf version)

pgkeller reviewed (changes requested):
ideally, this would be broken into multiple commits each w/ their own message.  that would leave a better history for blame and make it easier to bisect any possible regressions.  not required but preferred.

tt-aho committed:
#9012: Add handling of split commands for prefetch_d for CQ_PREFETCH_CMD_RELAY_PAGED_PACKED and CQ_PREFETCH_CMD_RELAY_INLINE_NOFLUSH

tt-aho committed:
#9012: Fix overlapping memory of ncrisc and profiler buffer l1 space.
Reduce L1_KERNEL_CONFIG_SIZE to 3KB
Note that this reduces available unreserved L1 space. (0.5KB on WH)

tt-aho committed:
#9012: Integrate new packed read/write cmds for FD
Add alignment parameter to CQDispatchWritePackedLargeCmd to allow for writing exact length, and striding through received data correctly
Add count parameter to CQPrefetchRelayPagedPackedCmd to allow for l1 to local cache copying of subcmds
Update tests to handle new params/cmd setup

tt-aho committed:
#9012: Add handling of split commands for exec buf for CQ_PREFETCH_CMD_RELAY_PAGED_PACKED and CQ_PREFETCH_CMD_RELAY_INLINE_NOFLUSH

tt-aho committed:
#0: Fix prefetch downstream ptr wrap case where local ptr was wrapped but the noc addr was not wrapped when downstream ptr is exexactly at the end of the buffer

tt-aho committed:
#0: Update metal Bert/Resnet perf targets after dispatch optimizations

TT-BrianLiu reviewed (approved):
Files owned by me looks good

tt-aho committed:
#9012: Fix l1_cache size to be correctly sized to l1_cache_elements_rounded
Update process_relay_paged_packed_sub_cmds to set last round value for write_pages_to_dispatcher to 1

tt-aho committed:
#0: Update microbenchmark thresholds after dispatch changes
